src/ShallowWaterExplicit.jl

```diff
@@ -0,0 +1,171 @@
+  # Forward integration of the shallow water equations using a supplied method
+  Ω = Triangulation(model)
+  dΩ = Measure(Ω, degree)
+  dω = Measure(Ω, degree, ReferenceDomain())
+  quad_cell_point = get_cell_points(dΩ.quad)
+  qₖ = Gridap.CellData.get_data(quad_cell_point)
+  wₖ = dΩ.quad.cell_weight
+  ξₖ = get_cell_map(Ω)
+
+  # Setup the trial and test spaces
+  reffe_rt  = ReferenceFE(raviart_thomas, Float64, order)
+  V = FESpace(model, reffe_rt ; conformity=:HDiv)
+  U = TrialFESpace(V)
+  reffe_lgn = ReferenceFE(lagrangian, Float64, order)
+  Q = FESpace(model, reffe_lgn; conformity=:L2)
+  P = TrialFESpace(Q)
+  reffe_lgn = ReferenceFE(lagrangian, Float64, order+1)
+  S = FESpace(model, reffe_lgn; conformity=:H1)
+  R = TrialFESpace(S)
+
+  # assemble the mass matrices
+  amm(a,b) = ∫(a⋅b)dΩ
+  H1MM = assemble_matrix(amm, R, S)
+  RTMM = assemble_matrix(amm, U, V)
+  L2MM = assemble_matrix(amm, P, Q)
+  H1MMchol = lu(H1MM)
+  RTMMchol = lu(RTMM)
+  L2MMchol = lu(L2MM)
+
+  # Project the initial conditions onto the trial spaces
+  b₁(q)   = ∫(q*h₀)dΩ
+  rhs1    = assemble_vector(b₁, Q)
+  hn      = FEFunction(Q, copy(rhs1))
+  ldiv!(L2MMchol, get_free_dof_values(hn))
+
+  b₂(v)   = ∫(v⋅u₀)dΩ
+  rhs2    = assemble_vector(b₂, V)
+  un      = FEFunction(V, copy(rhs2))
+  ldiv!(RTMMchol, get_free_dof_values(un))
+
+  b₃(s)   = ∫(s*f₀)*dΩ
+  rhs3    = assemble_vector(b₃, S)
+  f       = FEFunction(S, copy(rhs3))
+  ldiv!(H1MMchol, get_free_dof_values(f))
+
+  # initialise the diagnostics arrays
+  mass = Vector{Float64}(undef, nstep)
+  vort = Vector{Float64}(undef, nstep)
+  kin  = Vector{Float64}(undef, nstep)
+  pot  = Vector{Float64}(undef, nstep)
+  pow  = Vector{Float64}(undef, nstep)
+
+  # work arrays
+  h_tmp = copy(get_free_dof_values(hn))
+  w_tmp = copy(get_free_dof_values(f))
+
+  hm1    = FEFunction(Q, copy(get_free_dof_values(hn)))
+  um1    = FEFunction(V, copy(get_free_dof_values(un)))
+  hm2    = FEFunction(Q, copy(get_free_dof_values(hn)))
+  um2    = FEFunction(V, copy(get_free_dof_values(un)))
+  hp     = FEFunction(Q, copy(get_free_dof_values(hn)))
+  up     = FEFunction(V, copy(get_free_dof_values(un)))
+  ϕ      = FEFunction(Q, copy(get_free_dof_values(hn)))
+  F      = FEFunction(V, copy(get_free_dof_values(un)))
+  wn     = FEFunction(S, copy(get_free_dof_values(f)))
+  # first step, no leap frog integration
+  shallow_water_explicit_time_step!(model, dΩ, dω, f, g, hm1, um1, hm2, um2, hp, up, RTMMchol, L2MMchol, dt, false, τ, Q, V, R, S, hn, un, ϕ, F)
+  if mod(1, diag_freq) == 0
+    compute_diagnostics_shallow_water!(model, dΩ, dω, S, L2MM, H1MM, H1MMchol, h_tmp, w_tmp, g, hn, un, ϕ, F, mass, vort, kin, pot, pow, 1, true, wn)
+  end
+  
+  # subsequent steps, do leap frog integration (now that we have the state at two previous time levels)
+  for istep in 2:nstep
+    get_free_dof_values(hm2) .= get_free_dof_values(hm1)
```

Comment:
See this comment: https://github.com/gridapapps/GridapGeosciences.jl/pull/15#discussion_r700948954 

Is there any reason why you did not address it?

Reply:
When I tried to address this one it did not work (ie: the values did not change between time steps)

Reply:
That's interesting ... I do not understand how this cannot actually work. The only reason it cannot work is that the pointer that the FE functions hm2, hm1, etc. hold inside changes along the process, but I do not understand why is this the case in your code. 

Can you please add the following two lines at the beginning of the for loop:

```
println(objectid(get_free_dof_values(hm2)))
println(objectid(get_free_dof_values(hm1)))
```

Does the indentifier changes along the process? 

In any case, looking more carefully at this code, note that you are copying the entries. Is that really necessary? Can we instead use pointer swapping? (using the `FEFunction` constructor wisely)